### ISSUE TITLE
Feature Dependencies

### ISSUE THREAD

**Issue by virtualstaticvoid** [devcontainers/spec#90]
Hello,

I've written a couple of features for a project I'm working on and I've ran into situations where I'm not sure what the best approach to take is.

For example, I have a feature which involves installing a CLI tool using `npm`. Now, I'm not sure whether this feature should also install Node.js so that it is able to install the CLI tool, since I have no control over the base image, nor the included features in the `devcontainer.json` file.

The power of features, as I understand, is that they can be composed together; making each serve one purpose (installing the given feature) thus providing a simple way to handle all the tooling permutations for many different developers and/or teams whilst avoiding "golden container images"; and similar perhaps in concept to [buildpacks](https://buildpacks.io/).

So it occurred to me that since [`devcontainers/features`](https://github.com/devcontainers/features) already provide features for installing Node.js (and other language runtimes), it would be beneficial to make _my feature_ dependant on the `node` feature. Then I wouldn't need to worry about which base image a developer is using, and the `node` feature would be automatically included before mine, so my feature would install reliably.

Additional scenarios include:

* a Ruby gem which needs to build a native extension and thus requires `build-essential` to be installed.
* a Python package, installed via `pip`, needing to ensure Python and pip are installed.
* a Go CLI tool, which doesn't have a binary distribution, installed using `go install`, needing `go` to be installed.
* a `tflint` ruleset for AWS, which requires Terraform and `tflint` to be installed.

I realise installing packages using a `postCreateCommand` script can work, but it places the onus on the developer to know what features to include to install the given package. For example, I'm building features which install Python packages (e.g. `mkdocs`) for C# .Net developers, who wouldn't know to include the `python` feature.

I imagine a `devcontainer-feature.json` file including a `dependsOn` property, in the same schema as the `devcontainer.json` [`features` object](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#devcontainerjson-properties).

I.e.

```json
"dependsOn": {
  "ghcr.io/user/repo/go": {},
  "ghcr.io/user/repo1/go:1": {},
  "ghcr.io/user/repo2/go:latest": {},
  "https://github.com/user/repo/releases/devcontainer-feature-go.tgz": { 
        "optionA": "value" 
  },
  "./myGoFeature": { 
        "optionA": true,
        "optionB": "hello",
        "version" : "1.0.0"
  }
}
```

**Comment by joshspicer** [comment]
Thank you for your insights.  The extend in which we want to incorporate some notion of loose/strong dependencies (if any) has been a topic of discussion for a while now.  This is a valuable perspective.

The reference implementation and proposed spec provides an [`installsAfter`](https://containers.dev/implementors/features/#devcontainer-feature-json-properties) attribute that can be added to a feature's `devcontainer-feature.json`. 

`installsAfter` is an array of ID’s of Features that should execute before this one. Allows control for feature authors on soft dependencies between different Features

**Comment by virtualstaticvoid** [comment]
Ah, I wasn't aware of the `installsAfter` feature. Not sure how I missed that.

Thanks for the reply!

**Comment by joshspicer** [comment]
Please feel free to update us on whether `installsAfter` works for your use case. We would love the feedback (good and bad!)

**Comment by virtualstaticvoid** [comment]
Will do. I do have further feedback, related to installations into the user's home directory and also secrets, but I'll raise those with separate issues.

**Comment by virtualstaticvoid** [comment]
Has `installsAfter` been implemented in [`cli`](https://github.com/devcontainers/cli)?. 

I updated my feature JSON and it doesn't seem to work as expected.

Surprisingly, a [search](https://github.com/search?q=installsAfter&type=code) of Github for `installsAfter` contains very few results; and without any code which suggests it is implemented.

**Comment by joshspicer** [comment]
Yep, looks like a [small typo](https://github.com/search?q=repo%3Adevcontainers%2Fcli+installAfter&type=code) after some refactoring.  We'll be pushing changes shortly, and then afterwards please let us know your feedback :)
